src/placement/jump_map.c

```diff
@@ -381,13 +386,38 @@ get_target(struct pool_domain *curr_dom, struct pool_target **target,
 				continue;
 			}
 
+			/* Check if all domain range has been used for the current group */
+			range_set = isset_range(dom_cur_grp_used, start_dom, end_dom);
+			if (range_set) {
+				if (top == -1) {
+					/* all domains have been used by the current group,
+					 * then we cleanup the dom_cur_grp_used bits, i.e.
+					 * the shards within the same group might be put
+					 * to the same domain.
+					 */
+					memset(dom_cur_grp_used, 0, NBBY/dom_size + 1);
+					goto retry;
+				}
+				setbit(dom_cur_grp_used, curr_dom - root_pos);
+				curr_dom = dom_stack[top--];
+				continue;
+			}
+
 			range_set = isset_range(dom_used, start_dom, end_dom);
 			if (range_set) {
 				int idx;
 
 				/* Skip the domain whose targets are used up */
 				for (idx = start_dom; idx <= end_dom; ++idx) {
-					if (isclr(dom_occupied, idx))
+					/* Only reused the domain, if there are still targets
+					 * avaible (not being used) within this domain, and the
+					 * domain has not being used by current group yet. so
+					 * 1. there won't be multiple shards in the same target.
+					 * 2. there won't be multiple shards within same group
+					 *    are in the same domain.
+					 */
```

Comment:
What about SX object? And the isset_range() calls makes me worrying the performance.

Reply:
SX will only have 1 shard per group, what do you mean here?  

Actually the comments is a bit misleading, it actually means it will try not to put multiple shards within same group to be put into the same domain. But we do not promise that, if the group has been used all domain during rebuild, then we might still put multiple shards in the same domain.

Reply:
hmm, it should add much impacts for GX layout generation, might be some extra jump for some extrem corner case, but that is unavoided for GX anyway.